<?xml version="1.0" encoding="UTF-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
    <head>
        <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
        <title>My Hakyll Blog - Testing Faraday retries with WebMock</title>
        <link rel="stylesheet" type="text/css" href="../css/default.css" />
    </head>
    <body>
        <div id="header">
            <div id="logo">
                <a href="../">My Hakyll Blog</a>
            </div>
            <div id="navigation">
                <a href="../">Home</a>
                <a href="../about.html">About</a>
                <a href="../contact.html">Contact</a>
                <a href="../archive.html">Archive</a>
            </div>
        </div>

        <div id="content">
            <h1>Testing Faraday retries with WebMock</h1>

            <div class="info">
    Posted on June 20, 2018
    
</div>

<p>We use <a href="https://github.com/lostisland/faraday">Faraday</a> extensively for our making HTTP requests. We rely on <a href="https://github.com/bblimke/webmock">WebMock</a> to stub our these requests in rspec. We started encountering some timeout issues, and wanted to provide some retries without relying on a worker system. Fortunately, <strong>Faraday</strong> builds this in.</p>
<p>connection = Faraday.new do |config| config.request :retry, max: 2 end</p>
<p>Of course, I wanted to test this functionaltiy, which proved to be less than obvious. Fortunately, I discovered the <code>to_timeout</code> method stub. I was able to create the following two specs, which rely on the ability to chain methods in <strong>WebMock</strong> for subsequent calls.</p>
<p>before do class Foo def self.root response = connection.get(“/”) response.status end end end</p>
<p>context “retries twice” do it “succeeds after 2 timeouts” do stub_request(:get, stub_url) .to_timeout .to_timeout .to_return(status: 200)</p>
<p>expect { Foo.new.root }.not_to raise_error end</p>
<p>it “fails after 3 timeouts” do stub_request(:get, stub_url) .to_timeout .to_timeout .to_timeout .to_return(status: 200)</p>
<p>expect { Foo.new.root }.to raise_error end</p>
<p>It is nice to be able to test all aspects of our application, including configuration of a library.</p>

        </div>
        <div id="footer">
            Site proudly generated by
            <a href="http://jaspervdj.be/hakyll">Hakyll</a>
        </div>
    </body>
</html>
